File: packages/evaluators/octobot_evaluators/evaluators/channel/evaluators.py
# cython: language_level=3
#  Drakkar-Software OctoBot-Evaluators
#  Copyright (c) Drakkar-Software, All rights reserved.
#
#  This library is free software; you can redistribute it and/or
#  modify it under the terms of the GNU Lesser General Public
#  License as published by the Free Software Foundation; either
#  version 3.0 of the License, or (at your option) any later version.
#
#  This library is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
#  Lesser General Public License for more details.
#
#  You should have received a copy of the GNU Lesser General Public
#  License along with this library.

import async_channel.constants as channel_constants

import octobot_commons.logging as logging

import octobot_evaluators.evaluators.channel as evaluator_channels


class EvaluatorsChannelConsumer(evaluator_channels.EvaluatorChannelConsumer):
    """
    EvaluatorChannelConsumer adapted for EvaluatorsChannel
    """


class EvaluatorsChannelProducer(evaluator_channels.EvaluatorChannelProducer):
    """
    EvaluatorChannelProducer adapted for EvaluatorsChannel
    """

    # noinspection PyMethodOverriding
    async def send(self,
                   matrix_id,
                   data,
                   evaluator_name=channel_constants.CHANNEL_WILDCARD,
                   evaluator_type=channel_constants.CHANNEL_WILDCARD,
                   exchange_name=channel_constants.CHANNEL_WILDCARD,
                   cryptocurrency=channel_constants.CHANNEL_WILDCARD,
                   symbol=channel_constants.CHANNEL_WILDCARD,
                   time_frame=channel_constants.CHANNEL_WILDCARD,
                   origin_consumer=None):
        for consumer in self.channel.get_filtered_consumers(matrix_id=matrix_id,
                                                            evaluator_name=evaluator_name,
                                                            evaluator_type=evaluator_type,
                                                            exchange_name=exchange_name,
                                                            cryptocurrency=cryptocurrency,
                                                            symbol=symbol,
                                                            time_frame=time_frame,
                                                            origin_consumer=origin_consumer):
            await consumer.queue.put({
                "matrix_id": matrix_id,
                "evaluator_name": evaluator_name,
                "evaluator_type": evaluator_type,
                "exchange_name": exchange_name,
                "cryptocurrency": cryptocurrency,
                "symbol": symbol,
                "time_frame": time_frame,
                "data": data
            })


class EvaluatorsChannel(evaluator_channels.EvaluatorChannel):
    PRODUCER_CLASS = EvaluatorsChannelProducer
    CONSUMER_CLASS = EvaluatorsChannelConsumer

    MATRIX_ID_KEY = "matrix_id"
    EVALUATOR_NAME_KEY = "evaluator_name"
    EVALUATOR_TYPE_KEY = "evaluator_type"
    EXCHANGE_NAME_KEY = "exchange_name"
    CRYPTOCURRENCY_KEY = "cryptocurrency"
    SYMBOL_KEY = "symbol"
    TIME_FRAME_KEY = "time_frame"

    def __init__(self, matrix_id):
        super().__init__(matrix_id)
        self.logger = logging.get_logger(f"{self.__class__.__name__}")

    # noinspection PyMethodOverriding
    async def new_consumer(self,
                           callback: object,
                           size: int = 0,
                           priority_level: int = evaluator_channels.EvaluatorChannel.DEFAULT_PRIORITY_LEVEL,
                           matrix_id: str = channel_constants.CHANNEL_WILDCARD,
                           evaluator_name: str = channel_constants.CHANNEL_WILDCARD,
                           evaluator_type: object = channel_constants.CHANNEL_WILDCARD,
                           exchange_name: str = channel_constants.CHANNEL_WILDCARD,
                           cryptocurrency: str = channel_constants.CHANNEL_WILDCARD,
                           symbol: str = channel_constants.CHANNEL_WILDCARD,
                           time_frame=channel_constants.CHANNEL_WILDCARD) -> EvaluatorsChannelConsumer:
        consumer = EvaluatorsChannelConsumer(callback, size=size, priority_level=priority_level)
        await self._add_new_consumer_and_run(consumer,
                                             matrix_id=matrix_id,
                                             evaluator_name=evaluator_name,
                                             evaluator_type=evaluator_type,
                                             exchange_name=exchange_name,
                                             cryptocurrency=cryptocurrency,
                                             symbol=symbol,
                                             time_frame=time_frame)
        return consumer

    def get_filtered_consumers(self,
                               matrix_id=channel_constants.CHANNEL_WILDCARD,
                               evaluator_name=channel_constants.CHANNEL_WILDCARD,
                               evaluator_type=channel_constants.CHANNEL_WILDCARD,
                               exchange_name=channel_constants.CHANNEL_WILDCARD,
                               cryptocurrency=channel_constants.CHANNEL_WILDCARD,
                               symbol=channel_constants.CHANNEL_WILDCARD,
                               time_frame=channel_constants.CHANNEL_WILDCARD,
                               origin_consumer=None):
        return self.get_consumer_from_filters({
            self.MATRIX_ID_KEY: matrix_id,
            self.EVALUATOR_NAME_KEY: evaluator_name,
            self.EVALUATOR_TYPE_KEY: evaluator_type,
            self.EXCHANGE_NAME_KEY: exchange_name,
            self.CRYPTOCURRENCY_KEY: cryptocurrency,
            self.SYMBOL_KEY: symbol,
            self.TIME_FRAME_KEY: time_frame,
        },
            origin_consumer=origin_consumer)

    async def _add_new_consumer_and_run(self, consumer,
                                        matrix_id=channel_constants.CHANNEL_WILDCARD,
                                        evaluator_name=channel_constants.CHANNEL_WILDCARD,
                                        evaluator_type=channel_constants.CHANNEL_WILDCARD,
                                        exchange_name=channel_constants.CHANNEL_WILDCARD,
                                        cryptocurrency=channel_constants.CHANNEL_WILDCARD,
                                        symbol=channel_constants.CHANNEL_WILDCARD,
                                        time_frame=channel_constants.CHANNEL_WILDCARD):
        consumer_filters: dict = {
            self.MATRIX_ID_KEY: matrix_id,
            self.EVALUATOR_NAME_KEY: evaluator_name,
            self.EVALUATOR_TYPE_KEY: evaluator_type,
            self.EXCHANGE_NAME_KEY: exchange_name,
            self.CRYPTOCURRENCY_KEY: cryptocurrency,
            self.SYMBOL_KEY: symbol,
            self.TIME_FRAME_KEY: time_frame
        }

        self.add_new_consumer(consumer, consumer_filters)
        await consumer.run()

File: crates/octobot_evaluators/src/channels/evaluators.rs
use std::collections::HashMap;

use async_channel::constants::CHANNEL_WILDCARD;

use crate::channels::evaluator_channel::EvaluatorChannel;

// ---------------------------------------------------------------------------
// Filter key constants – mirror EvaluatorsChannel class attributes
// ---------------------------------------------------------------------------

pub const MATRIX_ID_KEY: &str = "matrix_id";
pub const EVALUATOR_NAME_KEY: &str = "evaluator_name";
pub const EVALUATOR_TYPE_KEY: &str = "evaluator_type";
pub const EXCHANGE_NAME_KEY: &str = "exchange_name";
pub const CRYPTOCURRENCY_KEY: &str = "cryptocurrency";
pub const SYMBOL_KEY: &str = "symbol";
pub const TIME_FRAME_KEY: &str = "time_frame";

// ---------------------------------------------------------------------------
// EvaluatorMessage
// ---------------------------------------------------------------------------

/// Typed message payload produced / consumed on an `EvaluatorsChannel`.
pub struct EvaluatorMessage {
    pub matrix_id: String,
    pub evaluator_name: String,
    pub evaluator_type: String,
    pub exchange_name: String,
    pub cryptocurrency: String,
    pub symbol: String,
    pub time_frame: String,
    pub data: HashMap<String, String>,
}

// ---------------------------------------------------------------------------
// EvaluatorsChannel
// ---------------------------------------------------------------------------

/// Channel dedicated to evaluator events. Wraps an `EvaluatorChannel` and
/// defines the set of filter keys used for routing.
/// Mirrors `evaluators/channel/evaluators.py`.
pub struct EvaluatorsChannel {
    pub inner: EvaluatorChannel,
}

impl EvaluatorsChannel {
    pub fn new(matrix_id: String) -> Self {
        Self {
            inner: EvaluatorChannel::new(matrix_id),
        }
    }
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

/// Build a filter map from the standard evaluator keyword arguments.
/// Any `None` value is replaced by `CHANNEL_WILDCARD`.
pub fn build_filter_map(
    matrix_id: Option<&str>,
    evaluator_name: Option<&str>,
    evaluator_type: Option<&str>,
    exchange_name: Option<&str>,
    cryptocurrency: Option<&str>,
    symbol: Option<&str>,
    time_frame: Option<&str>,
) -> HashMap<String, String> {
    let resolve = |v: Option<&str>| v.unwrap_or(CHANNEL_WILDCARD).to_string();
    let mut m = HashMap::with_capacity(7);
    m.insert(MATRIX_ID_KEY.to_string(), resolve(matrix_id));
    m.insert(EVALUATOR_NAME_KEY.to_string(), resolve(evaluator_name));
    m.insert(EVALUATOR_TYPE_KEY.to_string(), resolve(evaluator_type));
    m.insert(EXCHANGE_NAME_KEY.to_string(), resolve(exchange_name));
    m.insert(CRYPTOCURRENCY_KEY.to_string(), resolve(cryptocurrency));
    m.insert(SYMBOL_KEY.to_string(), resolve(symbol));
    m.insert(TIME_FRAME_KEY.to_string(), resolve(time_frame));
    m
}
